File: backend/src/routes/equipment.js
const { statusText } = require("../constants");
const { addAudit, machineById, nowText } = require("../domain");
const { parseJsonBody, requireFields, sendJson } = require("../http-utils");

async function handleEquipmentRoutes(req, res, pathname, store) {
  const equipmentStatusMatch = pathname.match(/^\/api\/equipment\/([^/]+)\/status$/);
  if (req.method === "POST" && equipmentStatusMatch) {
    const [, equipmentId] = equipmentStatusMatch;
    const body = await parseJsonBody(req);
    requireFields(body, ["status"]);
    const payload = await store.update((state) => {
      const machine = machineById(state, decodeURIComponent(equipmentId));
      if (!machine) {
        const error = new Error("Equipment not found");
        error.statusCode = 404;
        throw error;
      }
      machine.status = body.status;
      if (body.status === "alarm") {
        const alarmId = `ALM-${String(state.alarmSeq).padStart(3, "0")}`;
        state.alarmSeq += 1;
        state.alarms.unshift({
          id: alarmId,
          equipmentId: machine.id,
          severity: body.severity || "Medium",
          message: body.message || `${machine.name} 狀態異常，請確認機台訊號與保養紀錄。`,
          status: "alarm",
          createdAt: nowText()
        });
      }
      addAudit(state, `${machine.name} 狀態更新為 ${statusText[body.status] || body.status}`, body.actor || "Lab Operator");
      return { message: `${machine.name} 已更新` };
    });
    sendJson(res, 200, payload);
    return true;
  }

  return false;
}

module.exports = {
  handleEquipmentRoutes
};

File: backend/app/routes/equipment.py
from typing import Any
from urllib.parse import unquote

from fastapi import APIRouter, Request

from ..auth import require_roles
from ..errors import ApiError
from ..http_utils import read_json_body, require_fields
from ..services import equipment_service

router = APIRouter(prefix="/api/equipment", tags=["equipment"])


@router.post("/{equipment_id}/status")
async def change_equipment_status(equipment_id: str, request: Request) -> dict[str, Any]:
    require_roles(request.state.user, "operator")
    body = await read_json_body(request)
    require_fields(body, ["status"])
    return equipment_service.change_status(
        request.app.state.store,
        equipment_id=unquote(equipment_id),
        status=str(body["status"]),
        severity=str(body.get("severity") or ""),
        message=str(body.get("message") or ""),
        actor=str(body.get("actor") or "Lab Operator"),
    )


@router.put("/types")
async def configure_equipment_types(request: Request) -> dict[str, Any]:
    require_roles(request.state.user, "supervisor")
    body = await read_json_body(request)
    raw_types = body.get("types", [])
    if not isinstance(raw_types, list):
        raise ApiError("types must be an array", 400)
    return equipment_service.configure_types(
        request.app.state.store,
        types=raw_types,
        actor=str(body.get("actor") or "Lab Supervisor"),
    )
